src/refract/RenderFormURLEncodedVisitor.cc

```diff
@@ -0,0 +1,198 @@
+                    continue;
+                }
+
+                if ((*it)->element() == "ref") {
+                    HandleRefWhenFetchingMembers<T>(*it, members, IncludeMembers<T>);
+                    continue;
+                }
+
+                members.push_back(*it);
+            }
+        }
+    }
+
+    RenderFormURLEncodedVisitor::RenderFormURLEncodedVisitor(const std::string &key) : baseKey(key) {
+    }
+
+    RenderFormURLEncodedVisitor::~RenderFormURLEncodedVisitor() {
+    }
+
+    void RenderFormURLEncodedVisitor::operator()(const IElement& element) {
+        VisitBy(element, *this);
+    }
+
+    void RenderFormURLEncodedVisitor::operator()(const MemberElement& element) {
+        if (!element.value.first) {
+            return;
+        }
+
+        StringElement *keyElement = static_cast<StringElement*>(element.value.first);
+        if (!keyElement) {
+            return;
+        }
+
+        const StringElement::ValueType* key = GetValue<StringElement>(*keyElement);
+        if (!key) {
+            return;
+        }
+
+        std::string encodedKey = percentEncode(*key);
+
+        if (!result.str().empty()) {
+            result << "&";
+        }
+
+        if (element.value.second) {
+            if (!baseKey.empty()) {
+                encodedKey = baseKey + "[" + encodedKey + "]";
+            }
+
+            RenderFormURLEncodedVisitor visitor(encodedKey);
+            Visit(visitor, *element.value.second);
+            result << visitor.getString();
+        } else {
+            result << encodedKey << "=";
+        }
+    }
+
+    void RenderFormURLEncodedVisitor::operator()(const ObjectElement& element) {
+        ObjectElement::ValueType value;
+        IncludeMembers(element, value);
+
+        for (std::vector<refract::IElement*>::const_iterator it = value.begin();
+             it != value.end();
+             ++it) {
+
+            if (!*it) {
+                continue;
+            }
+
+            MemberElement *member = static_cast<MemberElement*>(*it);
+            if (member) {
+               operator()(*member);
+            }
+        }
+    }
+
+    void RenderFormURLEncodedVisitor::operator()(const EnumElement& element) {
+        const EnumElement::ValueType *value = GetEnumValue(element);
+
+        if (value && !value->empty()) {
+            operator()(*value->front());
+        }
+
+    }
+
+    void RenderFormURLEncodedVisitor::operator()(const ArrayElement& element) {
+        const ArrayElement::ValueType* value = GetValue<ArrayElement>(element);
+
+        if (value) {
+            for (ArrayElement::ValueType::const_iterator it = value->begin();
+                 it != value->end();
+                 ++it) {
+                if (!*it) {
+                    continue;
+                }
+
+                if (!result.str().empty()) {
+                    result << "&";
+                }
+
+                RenderFormURLEncodedVisitor visitor(baseKey + "[]");
```

Comment:
According to RFC 6570, the serialisation should be `list=green&list=red&list=blue`. No `[]` anywhere. If we have doubts regarding this, we can skip arrays for now. But, I would just remove the brackets.